webui/react/src/utils/role.ts

```diff
@@ -0,0 +1,121 @@
+import {
+  DetailedUser,
+  ModelItem,
+  ModelVersion,
+  Permission,
+  ProjectExperiment,
+  UserAssignment,
+  UserRole,
+  Workspace,
+} from 'types';
+
+// Permissions inside this workspace scope (no workspace = cluster-wide scope)
+const relevantPermissions = (
+  userAssignments?: UserAssignment[],
+  userRoles?: UserRole[],
+  workspaceId?: number,
+): Set<string> => {
+  if (!userAssignments || !userRoles) {
+    // console.error('missing UserAssignment or UserRole');
+    return new Set<string>();
+  }
+  const relevantAssigned = userAssignments.filter((a) => a.cluster ||
+    (workspaceId && a.workspaces && a.workspaces.includes(workspaceId))).map((a) => a.name);
+  let permissions = Array<Permission>();
+  userRoles.filter((r) => relevantAssigned.includes(r.name)).forEach((r) => {
+    // TODO: is it possible a role is assigned to this workspace,
+    // but not all of its permissions?
+    permissions = permissions.concat(r.permissions.filter((p) => p.globalOnly || workspaceId));
+  });
+  return new Set<string>(permissions.map((p) => p.name));
+};
+
+// Experiment actions
+export const canDeleteExperiment = (
```

Comment:
Could you make some kind of `usePermissions` hook so that users don't have to manually plumb data from the store (while also hiding the details of how you're storing permissions)?

Reply:
I imagine this would look like:
```typescript
const permissions = usePermissions()
permissions.canDeleteExperiment(experiment)
```

Reply:
I just got this working in a few parts, let me know if it makes sense to use this style in the other permissions. For example WorkspaceDetailsHeader

https://github.com/determined-ai/determined/pull/4965/files#diff-42d7ff039378e1e54f016def9813d3815bd8631e4caa3ef152d295c7ab34339a

and it goes through hooks/usePermissions

https://github.com/determined-ai/determined/pull/4965/files#diff-50e6e78f6aa3d324124d4e1d75ce89404b99f8e6e2aff107f58348d05d52a7a5R43

Reply:
Personally I think the experiment.ts stuff used in WorkspaceDetailsHeader should be hooks too since it's always using the same global data. That better encapsulates the intended functionality and I don't think it's a big rewrite. That said I'll leave this up to you.

Reply:
I added `usePermissions` for experiments, models, etc.
I removed several lines which pass the current user object from component to component, because we're now going to get the current user and check permission inside of `usePermissions`.

Reply:
If you can you should just make these functions private in the `usePermissions` hook file to reduce the surface area of the permissions "library".